src/main/java/mate/academy/EventManager.java

```
@@ -1,19 +1,76 @@
 package mate.academy;
 
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
 public class EventManager {
-    public void registerListener(EventListener listener) {
+    private final Set<EventListener> listeners = new CopyOnWriteArraySet<>();
+    private final ExecutorService executorService = Executors.newCachedThreadPool();
+    private final AtomicBoolean isShutdown = new AtomicBoolean(false);
 
+    public void registerListener(EventListener listener) {
+        if (listener == null) {
+            throw new NullPointerException("Listener can't be null");
+        }
+        if (isShutdown.get()) {
+            throw new IllegalStateException("EventManager is shut down");
+        }
+        listeners.add(listener);
     }
 
     public void deregisterListener(EventListener listener) {
-
+        if (listener != null) {
+            listeners.remove(listener);
+        }
     }
 
     public void notifyEvent(Event event) {
+        if (event == null) {
+            throw new IllegalArgumentException("Event cannot be null");
+        }
+        if (isShutdown.get()) {
+            throw new IllegalStateException("EventManager is already shut down");
+        }
 
+        List<EventListener> currentListeners = List.copyOf(listeners);
+        for (EventListener listener : currentListeners) {
+            try {
+                executorService.submit(() -> {
+                    try {
+                        listener.onEvent(event);
+                    } catch (Exception e) {
+                        e.printStackTrace();
+                    }
+                });
+            } catch (Exception e) {
+                System.err.println("Task rejected: " + e.getMessage());
+            }
```

You check `isShutdown.get()` before scheduling tasks, but that check is not atomic with calls to `executorService.submit(...)`. If `shutdown()` runs between the check and `submit`, submissions may be rejected. Instead of catching all `Exception` around `submit`, explicitly catch `RejectedExecutionException` and handle it (skip scheduling that listener and optionally log). This documents the intended behavior when shutdown races with notifyEvent and avoids swallowing unrelated exceptions silently.